### PR TITLE
[4/N] Pack-first DP schedule; global_batch_size counts rollouts

### PR DESCRIPTION
**Pull request by Zhichenzzz**
## Summary

Stacked on #1931. Parts C+E of slime #1933 — the payoff PR of the series: the schedule packs first and distributes second, and `global_batch_size` counts ROLLOUTS instead of training samples, so compact / subagent rollouts (one execution -> N training samples) train correctly.

- `build_dp_schedule` rewritten pack-first-distribute-second: group samples by rollout id, split rollouts into steps of `global_batch_size` rollouts, pack each step's samples into `K` mbs with one global first-fit pass, align `K` to `dp_size * mb_group`, then distribute mbs across ranks (Karmarkar-Karp on mbs token sums under `balance_data`, strided otherwise). Compact rollouts stay whole within one training step; trailing rollouts that don't fill a step are dropped whole.
- Postprocess skips the sample-count trim when compact rollouts are detected (explicit `rollout_id`) — cutting mid-rollout would orphan siblings.
- `log_rollout_data` per-rollout-mean metrics count rollouts (`num_rollouts / dp`) for scheduled shards.
- New `miles/rollout/generate_hub/compact_split_example.py`: minimal compact generate fn (splits each response into two training samples sharing the parent's `rollout_id`) used for e2e validation of the whole pipeline.

Deviation from slime #1933: `--use-dynamic-global-batch-size` / `--disable-rollout-trim-samples` are kept — multi-LoRA and the legacy backend paths still use them; only the scheduled megatron path adopts the by-rollout semantics.

## Validation

- CPU: 97 tests green — schedule invariants incl. compact grouping (`test_rollout_grouping_keeps_samples_together`), whole-rollout trailing trim, randomized sweep with random rollout groupings, static misalignment assertion, plus the full PR2-PR4 batteries.
- GPU e2e on 8xH200 (`--ci-test`):

| run | topology | result |
|---|---|---|
| Qwen2.5-0.5B gsm8k short | dp8, dynamic batch, mooncake | ✅ succeeded |
| Qwen3-30B-A3B baseline | dp2 + cp2 + pp2 + ep2, metric gates | ✅ succeeded, `logprob_abs_diff` 0.0176 (branch series 0.0171-0.0178, same topology — no-op confirmed for 1 rollout = 1 sample) |
| Qwen3.5-35B-A3B (GDN) | cp2 | ✅ succeeded (see thread) |
| **compact gain demo** | 0.5B dp8 + `compact_split_example` | ✅ succeeded |

### Compact gain evidence

With the example splitter (each rollout -> 2 training samples), the schedule log reads:

```
Rollout-side DP schedule: num_samples=64, num_rollouts=32, global_batch_sizes=[32], num_microbatches=[1]
```

i.e. 32 rollout executions producing 64 training samples train as ONE step of 32 rollouts with per-rollout loss weighting. Before this series the same workload either failed validation or was treated as 64 independent samples: two training steps per rollout batch, siblings split across steps/trim boundaries, and every multi-sample rollout over-weighted in the loss proportionally to its sample count. Training over 3 rollout batches is stable (rewards 0.44-0.56, healthy logprob alignment).

### Stress test

Extreme variant (`generate_stress`, devbox-only): each rollout splits into `1 + (index % 8)` siblings (1-8, variable lengths), under **cp2** + dynamic batching with a small `--max-tokens-per-gpu 2048` so siblings scatter across micro-batches AND DP ranks, with `check_for_nan_in_loss_and_grad` on:

```
Rollout-side DP schedule: num_samples=288, num_rollouts=64, global_batch_sizes=[32, 32], num_microbatches=[3, 3]
```

64 rollout executions -> 288 training samples, still 2 steps of 32 rollouts; cross-mb/cross-rank per-rollout denominators reconstruct correctly (no NaN, stable rewards 0.52-0.62 over 3 rollout batches, job succeeded).

### CODE REVIEW

**Comment by Zhichenzzz** [comment]
Consolidated into #1929 (the stack is now #1927 -> #1929).
